app/models/sms_message.rb

```diff
@@ -52,4 +52,70 @@ class SmsMessage < ApplicationRecord
                   received: "received",
                   read: "read",
   }, prefix: :status
+
+  def self.save_api_response(message, recipient)
+    retries ||= 0
+    record = find_or_initialize_by(message_sid: message.sid)
+
+    create_only_fields = {
+      body: message.body,
+      recipient: recipient
+    }
+
+    create_or_update_fields = {
+      from: message.from,
+      to: message.to,
+      date_created: message.date_created,
+      date_updated: message.date_updated,
+      date_sent: message.date_sent,
+      direction: message.direction,
+      error_code: message.error_code,
+      error_message: message.error_message,
+      status: message.status
+    }
+
+    if record.persisted? && !has_updated_status?(record.status, message.status)
+      record.update(create_only_fields) if message.body
+      return record
+    end
+
+    record.update(create_or_update_fields.merge(create_only_fields))
+    record
+  rescue ActiveRecord::RecordNotUnique
+    retries += 1
+    retry if retries <= 1
+  end
+
+  def self.save_status_callback(params)
+    retries ||= 0
+    record = find_or_initialize_by(message_sid: params['MessageSid'])
+
+    return record if record.persisted? && !has_updated_status?(record.status, params['MessageStatus'])
+
+    record.update(
+      error_code: params['ErrorCode'],
+      error_message: params['ErrorMessage'],
+      from: params['From'],
+      to: params['To'],
+      status: params['MessageStatus']
+    )
+    record
+  rescue ActiveRecord::RecordNotUnique
+    retries += 1
+    retry if retries <= 1
+  end
+
+  private
+
+  def self.has_updated_status?(old_status, new_status)
+    new_index = status_index(new_status)
+    old_index = status_index(old_status)
+    return true unless old_index
+    return unless new_index
```

Comment:
No, but you've inspired me to be more explicit than gcf and say `return false` here instead of just `return`. If there is no new_index (ie `unless new_index`), then `new_status` is either nil or isn't in the list of statuses, and therefore it is not an updated status.